HW5.py

```diff
@@ -0,0 +1,586 @@
+    )
+
+    return file_name
+
+
+def plot_difference_with_mean_weighted(
+    x1, x2, y, num_bins, pred1_name, pred2_name, pred1_type, pred2_type
+):
+    file_name = f"diffmeanplots/DiffmeanplotWeighted - {pred1_name} - {pred2_name}.html"
+    if pred1_type == "cat":
+        x1, _ = pd.factorize(x1)
+    if pred2_type == "cat":
+        x2, _ = pd.factorize(x2)
+    hist, xbins, ybins = np.histogram2d(x1, x2, bins=num_bins)
+    _, x1bins = np.histogram(x1, num_bins)
+    _, x2bins = np.histogram(x2, num_bins)
+    bin_means, binx_edges, biny_edges, binnumber = binned_statistic_2d(
+        x1, x2, y, statistic="mean", bins=[x1bins, x2bins], range=None
+    )
+    population_mean = np.average(y)
+    pop_proportion = hist / len(y)
+    mean_squared_diff = np.power((bin_means - population_mean), 2)
+
+    mean_squared_diff_weighted = mean_squared_diff * pop_proportion
+    fig = px.imshow(mean_squared_diff_weighted)
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+    return file_name
+
+
+def difference_with_mean_2d(
+    x1, x2, y, num_bins, pred1_name, pred2_name, pred1_type, pred2_type
+):
+    if pred1_type == "cat":
+        x1, _ = pd.factorize(x1)
+    if pred2_type == "cat":
+        x2, _ = pd.factorize(x2)
+    hist, xbins, ybins = np.histogram2d(x1, x2, bins=num_bins)
+    _, x1bins = np.histogram(x1, num_bins)
+    _, x2bins = np.histogram(x2, num_bins)
+    bin_means, binx_edges, biny_edges, binnumber = binned_statistic_2d(
+        x1, x2, y, statistic="mean", bins=[x1bins, x2bins], range=None
+    )
+    population_mean = np.average(y)
+    pop_proportion = hist / len(y)
+    mean_squared_diff = np.power((bin_means - population_mean), 2)
+    mean_squared_diff_sum = np.nansum(mean_squared_diff)
+    mean_squared_diff_weighted = mean_squared_diff * pop_proportion
+    mean_squared_diff_weighted_sum = np.nansum(mean_squared_diff_weighted)
+    unweighted_plot = plot_difference_with_mean_unweighted(
+        x1, x2, y, 10, pred1_name, pred2_name, pred1_type, pred2_type
+    )
+    weighted_plot = plot_difference_with_mean_weighted(
+        x1, x2, y, 10, pred1_name, pred2_name, pred1_type, pred2_type
+    )
+    df = pd.DataFrame(
+        list(
+            zip(
+                [pred1_name],
+                [pred2_name],
+                [pred1_type],
+                [pred2_type],
+                [population_mean],
+                [mean_squared_diff_sum],
+                [mean_squared_diff_weighted_sum],
+                [unweighted_plot],
+                [weighted_plot],
+            )
+        ),
+        columns=[
+            "Predictor 1",
+            "Predictor 2",
+            "Predictor 1 Type",
+            "Predictor 2 Type",
+            "PopulationMean",
+            "MeanSquaredDiff",
+            "MeanSquaredDiffWeighted",
+            "Unweighted Plot",
+            "Weighted Plot",
+        ],
+    )
+    return df
+
+
+def fun(path):
```

Comment:
If you have time, you may want to add hyperlinks on your predictors that link to the plots we did in HW4 (I think this was one of the requirements).